### ISSUE TITLE
Incorrect alpha compositing using "matplotlib.pyplot.imshow".

### ISSUE THREAD

**Issue by KelSolaar** [matplotlib/matplotlib#9906]
### Bug report

**Bug summary**

`matplotlib.pyplot.imshow` definition alpha compositing is incorrect, references #3343.

**Code for reproduction**

```python
from matplotlib import pyplot as plt

plt.figure(figsize=(8, 8))
RGBA = plt.imread('RGB_Droplet.png')
plt.imshow(RGBA)
plt.show()
```

**Actual outcome**

![image](https://user-images.githubusercontent.com/99779/33501848-f2097b12-d742-11e7-9bd2-824ce72d2cab.png)

**Expected outcome**

Composited in The Foundry Nuke:

![image](https://user-images.githubusercontent.com/99779/33501902-1a3bcd4c-d743-11e7-953e-968ebbd3eec9.png)

The images used are available here:

- [RGB_Droplet.png](https://drive.google.com/file/d/1Ct7o2TPWc1giFWSTYa4O0oyRljSWihIO/view?usp=sharing)
- [RGB_Droplet_Over_White_Reference.png](https://drive.google.com/file/d/1uIT496wEz1pHmqcLYdlu2Cq5OkzQBjUQ/view?usp=sharing)

**Matplotlib version**
  * Operating system: `macOS`
  * Matplotlib version: `'2.1.0'`
  * Matplotlib backend (`print(matplotlib.get_backend())`): `'module://ipykernel.pylab.backend_inline'`
  * Python version: `'2.7.13 |Continuum Analytics, Inc.| (default, Dec 20 2016, 23:05:08) \n[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)]'`
  * Jupyter version (if applicable): `notebook                  5.0.0                    py27_0`

**Comment by anntzer** [comment]
Possibly related to https://github.com/matplotlib/matplotlib/issues/8847?

**Comment by QuLogic** [comment]
Here's how it looks in the Drive preview (albeit with checkerboard background):
![screenshot-2017-12-5 rgb_droplet png](https://user-images.githubusercontent.com/302469/33588424-bcba1fc0-d940-11e7-8eb0-472043ed4b86.png)
and here's how it looks in eog with white background:
![screenshot from 2017-12-05 01-04-22](https://user-images.githubusercontent.com/302469/33592311-4db40a10-d958-11e7-84ca-2f04fac97f07.png)
or in gimp with a white layer below it:
![screenshot from 2017-12-05 01-09-33](https://user-images.githubusercontent.com/302469/33592433-fa83398c-d958-11e7-9c00-6f87bff488e3.png)

So I'm not sure this is a bug per se. It is possible that one is using premultiplied alpha and the other is not (when the incoming image is the opposite.)

**Comment by KelSolaar** [comment]
My image has alpha premultiplied, so I guess what Matplotlib doing is multiplying again RGB by alpha, this should probably be an optional behavior.

**Comment by KelSolaar** [comment]
Seems to be it:

![image](https://user-images.githubusercontent.com/99779/33593466-76d67030-d9f4-11e7-8d05-05b8f31dd79e.png)

Using:

[RGB_Droplet_UnPremultiplied.png](https://drive.google.com/file/d/17kXLJTszI92T28btk7E4KM3wkfBOs37i/view?usp=sharing)

**Comment by efiring** [comment]
It's not clear to me whether there is a proposed change for mpl here.  If not, this should be closed.  If there is something that needs to be done, perhaps it should be stated at the top of a new issue, with a reference to this one, and then close this one.

**Comment by KelSolaar** [comment]
I don't know if it is at image read time or display time, but MPL should offer an option for different behaviour whether the RGB channel has alpha pre-multiplied or not. Currently the RGB channel is getting multiplied potentially two times, so this is an issue.

- If the RGB channels are already pre-multiplied by alpha, you don't multiply.
- If the RGB channels are not pre-multiplied by alpha, you multiple by alpha.

**Comment by efiring** [comment]
I think the option would have to be at image read time; internally, mpl has to consistently use one form or the other, and presently it is using straight RGBA.  It seems like the underlying problem here is that RGBA is not well-defined.  Therefore imread could gain a kwarg, defaulting to `premultiplied=False`; when set to True it would cause the input RGB values to be divided by the alpha value.  Is this correct?  If so, would you like to submit a PR for this?

**Comment by jklymak** [comment]
PNG spec says not to premultiply alpha, so I don't think we are doing anything incorrect by reading it that way: https://www.w3.org/TR/PNG-Rationale.html  If the user wants to divide by alpha after reading, they are welcome to do so.

**Comment by KelSolaar** [comment]
Just to be clear, I have the same issue if I pass an image generated directly by code to `imshow`.

I could not find anywhere in MPL `imshow` documentation a statement that image must be either premultiplied or un-premultiplied, this should be clarified.

**Comment by jklymak** [comment]
I'm not following your first statement.  Do you have an example?  

I think the documentation is pretty clear.  If you want to specify RGB, you pass RGB, if you want to specify RGBA, you pass RGBA.  I'm not clear why you would you pass RGBA to `imshow` if you've already applied the alpha to the RGB data. What would `imshow` do with the A channel?

**Comment by KelSolaar** [comment]
The question is not whether you pass alpha or not, but whether your RGB has been premultiplied with Alpha or not. The capability of doing A over B through an alpha is totally independent of the state of RGB.

**Comment by KelSolaar** [comment]
On the left, `A`, i.e. the RGB droplet is already premultiplied by `Mask`, i.e. the black and white droplet and is composited over the `B`, i.e. the grid background using `Mask` as an alpha.

On the righ, `A`, i.e. the RGB droplet is already premultiplied by `Mask`, i.e. the black and white droplet and is divided by `Mask` and is composited over the `B`, i.e. the grid background using `Mask` as an alpha.

As you can see A over B is possible in both cases, the problem is that depending whether your RGB values have been premultiplied, you have to account for it, and thus at the very least, indicate clearly in the document what MPL expect, which seems to be un-premultiplied RGB data to be sent to `imshow`.

![image](https://user-images.githubusercontent.com/99779/33738345-e5716bca-dbfd-11e7-8b45-a93d693b4fb4.png)

**Comment by jklymak** [comment]
Matplotlib `imshow` has no concept of premultiplied.  It takes an MxNx3 RGB array of floats and displays those values.  Matplotlib doesn't care how those values were produced.  If you pass it an MxNx4 RGBA array it displays those RGB values with an alpha given by A.  The image is displayed with the alpha mixed with *whatever* background exists on the figure behind the image.  

The image you supplied has already had A applied against *black* background.  Passing A again, as you are doing above, masks the black background, but of course still allows some of the side data to leak in and look greyish.  

I really don't think this is a bug, and I really think the documentation is clear.  

That there is a community out there that pre-multiplies alpha into their images and then still supplies that alpha in the PNG isn't really our business, but in my opinion, they are going against the spec and missing the point of having an alpha channel.  

![fig](https://user-images.githubusercontent.com/1562854/33738709-087438fa-db4f-11e7-8535-9dba92204f57.png)

```python

from matplotlib import pyplot as plt
import numpy as np

fig, axs = plt.subplots(1, 2,)
RGBA = plt.imread('RGB_Droplet.png')
axs[0].imshow(RGBA)
axs[1].imshow(RGBA[:,:,:-1])
plt.show()
```

**Comment by KelSolaar** [comment]
PNG is probably not the good file format to discuss about since the spec seems to advocate straight / un-premultiplied alpha.

> I really think the documentation is clear.

I would argue the opposite considering:

1.  The fact people have asked on Stackoverflow: https://stackoverflow.com/questions/25089068/how-does-imshow-handle-the-alpha-channel-with-an-m-x-n-x-4-input
2. There is a related issue opened here: https://github.com/matplotlib/matplotlib/issues/3343
3. The very issue we are discussing on: https://github.com/matplotlib/matplotlib/issues/9906

I don't think a mention to straight un-premultiplied and premultiplied alpha compositing, matplotlib behavior in that regard along a possible reference to https://en.wikipedia.org/wiki/Alpha_compositing wouldn't kill in the `imshow` documentation.

I have scratched my head a few hours trying to understand what is happening and I'm not alone.

**Comment by jklymak** [comment]
#3343 and the S/O question are the same thing and both were fixed by #5718.  If you run the code in #5718 you get the expected result (all white).  The state of affairs before #5718 was indeed incorrect and a bug.  

Please provide a minimal self-contained example that does not require opening an image that displays the problem you think still exists.

**Comment by KelSolaar** [comment]
As of this writing I'm leaning on the side that the documentation is scarce in term of describing what is the alpha compositing process in MPL not so much on there is an issue. With the myriad of image manipulation packages, image file formats and way of writing those images, I think MPL should describe its alpha compositing process. As a matter of fact, `imshow` documentation does not describe how it performs its over operation, it is a black box and one would have to dive into the code to understand it.

**Comment by jklymak** [comment]
If alpha is not documented anywhere, then that is indeed an oversight.   

99% sure that its just: RGB_back is the background and RGB_fore is the foreground, its `RGB = RBG_back * (1 - alpha) + RGB_fore * alpha`.  The `zorder` sets what is the foreground.

**Comment by KelSolaar** [comment]
Thanks @jklymak!

**Comment by jklymak** [comment]
No problem - I know next to nothing about "formal" image processing, so I only have the "basic" intuition about what alpha could mean.  Fortunately for me, I think thats the same as how Matplotlib thinks about it. You've obviously used it in different contexts, so are approaching it from a different point of view.  Its definitely helpful to get different points of view so we can make the docs as universal as possible.  

I'd suggest https://matplotlib.org/tutorials/colors/colors.html#sphx-glr-tutorials-colors-colors-py as a decent place to put the appropriate definitions, though your suggestion as to where else it should go would be appreciated.  However, bear in mind that Matplotlib uses alpha for most objects, so we aren't going to explain it for every function call!

Thanks!

**Comment by anntzer** [comment]
Closed by #9959 and #9981, afaict.
